<a href="https://colab.research.google.com/github/ravichas/bifx-546/blob/main/Notebooks/Chapter7_HypTesting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Chapter 7 Hypothesis Testing

Often Data Scientists want to test whether a certain hypothesis is likely to be true.


Hypothesis: A claim that we want to test

**Null Hypothesis** (H$_0$)

* Currently accepted value for a parameter (ex., mean)

**Alternate Hypothesis** (H$_a$) or Research Hypothesis

* Involves the claim to be tested

Another way to look at this is

Hypthesis Testing:

* Formal method DS use to decide whether a claim about the world is supported by data or not

 **The Process**
  * **Start with a claim you want to test **
    * Example: "The coin is fair" (50/50 chance of H/T)
  * A**ssume the claim is true, and figure out what you would expect to see **
    * if the coin is fair, flipping it 1000 times should give you 500 Hs (may be 48, may be 53, but somewhere close)
    * We know the pattern of what "randomness" looks like when the coin is fair
  * **Collect actual data**
    * Flip the coin 1000 times and count: you got 991 Hs
  * **Ask: "How weird is what I actually saw?"**
    * If the coin were fair, getting 900 heads out of 1000 would be extremely unlikely
    * This suggests our original assumption (fair coin) is probably wrong
  * **Make a decision**
    * If your data looks normal/expected --> The HYP is probably true
    * If your data looks weird/unlikely --> the HYP is probably false.


According to Tibshirani et al book on Statistical Learning, Hyp Testing is a rigorous stat framework for answering simple "yes-no" questions about the data

Example

We operate a candy-bar manufacturing plant and recently upgraded our production machines. The factory quality engineer now claims that the average weight of our candy bars is no longer 250 grams. To evaluate this claim, we set up the following hypotheses:


$$ H_0: \mu = 250 \quad \text{(The factory still produces candy bars with a mean weight of 250 grams.)} $$ (default belief)

$$ H_a: \mu \ne 250 \quad \text{(The mean weight has changed from 250 grams after the upgrades.)} $$

Note that $H_{0}$ and $H_{a}$ are mathematical opposites. For example, $H_a$ is whatever that $H_0$ is not.

We assume that the null is true unless the data proves otherwise.

Let us explore the possible outcomes for this test:
* Reject Null Hyp
* Fail to **Reject Null Hypothesis** (difficult to prove something is true)

Think of US judicial system while reading through this.


# How can we carry out the test?

* First we need a $H_0$ (default belief; nothing has changed after the upgrade) and $H_a$. Note $H_0$ is the baseline assumption.
* We need something called Test Statistic (that provides evidence for rejecting the $H_0$. This is calculated from sample data and is used to either reject the $H_0$ or **Fail** to reject the $H_0$
$$ T = \frac{\bar{x} - 250 } {s/\sqrt{n}} $$
Where:
  * $\bar{x}$ = sample mean weight
  * s = samp SD
  * n = sample size
Note:
  * T = 0 --> The sample mean equals 250 exactly --> Consistent with $H_0$
  * T large postive: sample mean is far above 250
  * T is large negative:  sample mean is far below 250
  * large |T| means stronger evidence against $H_0$. The test statistic summarizes how consistent the data are with $H_0$.
* We then compute a $p-value$ that measures the probability of having observed a comparable or extrement value of the **test statistic** under the $H_0$. In essence, p-value answers:
  * **If the true mean is 250 gms ($H_0$ is true), how unusual is the sample we observed?**
  * Note, If the null hypothesis holds, the test statistic T approximately follows a t-distribution (or approximately N(0,1) for large n).
  * Note: "$H_0$ holds" means:
    * The true weight of all candy bars produced by the factory (entire population) is actually 250 grams.
* Finally using the p-value we either "reject the $H_0$" or "Fail to reject" $H_0$
  * Let us consider the example: If $T=233$. This means that our sample mean is 2.33 SD away from 250 grams. This value in the distribution lies at the extreme and lies outside of $\pm$ 2.33 that means only 2% of samples would give such an extreme result if $\mu = 250$ were actually true. Hence, p=0.02 (one-sided) and p=0.02*2 = 0.04
  * We need to decide whether 4% small? We also have a pre-determined cutoff ($\alpha$). Using $\alpha$, we Reject $H_0$ if p $\lt$ 0.05, Fail to reject if $H_0$ if p >= 0.05. Based on this, we reject the null-hypothesis at $alpa$ = 0.05

# Errors:

* Type 1 Error (False Postive): $H_0$ is actually true (true mean is still 250g). But, we reject $H_0$. We conclude the machine changed the mean weight even though it did not.
  * Consequences:
    * We stop production
    * we recalibrate the machine
    * blame the upgrade
    * but nothing was wrong
  * Are Type-1 errors bad?
* Type II Error (False Negative): $H_{0}$ is actually false (mean has really changed). But, we failed to reject the Null.  
    * Consequences:
        * We continue production without noticing
        * Candy. bars size over/under-filled
        * Customers may complain
        * Quality issues arise
      * Are Type-II errors bad?
* Power of the test
  * Power = 1 - Type II error rate
    * Our test correctly detects real changes
* Trade-off between Type1/TypeII errors and the choice of $\alpha$ ?
  * You cannot minimize both errors simultaneously
    * Lowering Type I error increase Type II error
    * Lowering Type II increase Type I error

# Statistically Significant  (SS)

A result is SS when

* p-value $\lt$ $\alpha$

Note:
* SS does not mean
  * effect is big
  * result is true
  * more important

# Example of testing whether the coin fair

if we want to test whether the coin (p = prob of landing Hs) we have is fair.  

$$ H_0 : p = 0.5 $$
$$ H_A : p \ne 0.5 $$

# Basic ideas

* Bernoulli Trial (Single flip)
  * Each individuval flip is a "Bernoulli trial"
  * Only two outcomes: H (Sucess) or T (failure)
  * Probability of Hs = p (for a fair coin, p = 0.5)
* Binomial Distribution (Multiple Independent Flips)
  * A reference table/curve that tells you probabilities
  * Created mathematically (not from your experiemtn)
  * Shows what outcomes to expect IF the coin is fair (p=0.5)
  ### Connection to Bernoulli
    * Each of the flip is a Bernoulli trial
    * X = number of Hs in one trial
    * X follows a Binomial(n=100, p=0.5) distribution
  ### What Binomial Distribution Tells You:
    * P(X=0) = Prob of 0 Hs in 100 flips
    * ...
    * P(X=65) = Prob of 65 Hs in 100 flips
    * etc.

  $$ P(X = k) = C(n, k) \times p^{k} \times (1-p)^{(n-k)} $$
  where $C(n,k) = "n choose k" = combinations

* Normal Approximation (when n is large; number of times to repeat the bernoulli trial)



# Real Experiment
  * Flip a coin = 100 times
  * Cont the result, X = 65 Hs (example)
  * One experiment. One Number: 65
* Question
  * Is this coin fair, or is it biased?

# Using a pre-existing Reference Distribution

  * Mathematicians already derived a reference distribution for us
    * They derived the Binomial(n=100, p=0.5) distribution
    * This distribution shows, "If the coin is fair, here is likely each outcome is"
    * It tells us the probability of getting 0 Hs, 1Hs, ....100 Hs.
  * If we take one result from our experiment (X = 65) and compare it with the reference
    * Look at our pre-existing Binomial Distribution
    * Ask: Where does my result (H=65) fall on this distribution?
    * Find: "Getting 65 or more Hs happens only 0.3% of the time if the coin is fair
  * Make a decision
    * Since 65 is unlikely under the "fair coin" distribution
    * Conclude: The coin is biased


# Mean

$$ E[[X] = \sum k \times P(X=k)  \text{   for k = 0 to n} $$

After working with Algebra, we get

E[X] = np

Example

* If we flip 100 coin (n = 100)
* Each has p = 0.5 chance of Hs
* Expected Hs $\mu$ = 100 * 0.5 = 50

# SD

$$ \text{Var}[X] = E[X^2] - (E[X])^2 $$

After working through Algebra,

$$ \text{Var}[X] = \sqrt{np(1-p)} $$

# Example

* n= 100, p = 0.5
* Variance = 100 * 0.5 * 0.5 = 25
* SD: $\sigma$ = $\sqrt{25}$ = 5



As n increases, Binomial(n, p) --> Normal Distribution

The limiting normal has the SAME mean (np) and variance (np(1-p))

When np$\ge$10 and n(1-p) $\ge$ 10, he approximation is accurate enough for practical use.

In [5]:
# ==== Course setup: safe clone + cd + import path ====
import os
import sys

REPO_URL = "https://github.com/joelgrus/data-science-from-scratch.git"
REPO_DIR = "data-science-from-scratch"

# 1. If we're *anywhere inside* the repo, move to the parent directory first
cwd = os.getcwd()
if REPO_DIR in cwd.split(os.sep):
    parts = cwd.split(os.sep)
    # Walk up until we are at .../data-science-from-scratch
    while parts and parts[-1] != REPO_DIR:
        parts.pop()
    # Now go to the directory *above* the repo
    parent_dir = os.sep.join(parts[:-1]) or "/"
    os.chdir(parent_dir)
    print(f"Moved to parent directory: {os.getcwd()}")

# 2. Clone only if needed
if not os.path.exists(REPO_DIR):
    print("Cloning repo...")
    !git clone {REPO_URL}
else:
    print(f"{REPO_DIR} already exists — skipping clone.")

# 3. cd into the repo (this is where you'll live most of the time)
%cd {REPO_DIR}

Cloning repo...
Cloning into 'data-science-from-scratch'...
remote: Enumerating objects: 392, done.
remote: Counting objects: 100% (145/145), done.
remote: Compressing objects: 100% (52/52), done.
remote: Total 392 (delta 107), reused 93 (delta 93), pack-reused 247 (from 1)
Receiving objects: 100% (392/392), 762.39 KiB | 4.68 MiB/s, done.
Resolving deltas: 100% (207/207), done.
/content/data-science-from-scratch


In [6]:
from typing import Tuple
import math

def normal_approximation_to_binomial(n: int, p: float) -> Tuple[float, float]:
  """ Returns mu and sigma corresponding to a Binomial(n, p) """
  mu = p * n
  sigma = math.sqrt(p * (1-p) * n)
  return mu, sigma


In [7]:
from scratch.probability import normal_cdf
# The normal cdf _is_ the probability the variable is below a threshold
normal_probability_below = normal_cdf
# It's above the threshold if it's not below the threshold
def normal_probability_above(lo: float, mu: float = 0, sigma: float = 1) -> float:
  """The probability that an N(mu, sigma) is greater than lo."""
  return 1 - normal_cdf(lo, mu, sigma)

# It's between if it's less than hi, but not less than lo
def normal_probability_between(lo: float, hi: float, mu: float = 0,
                               sigma: float = 1) -> float:
  """The probability that an N(mu, sigma) is between lo and hi."""
  return normal_cdf(hi, mu, sigma) - normal_cdf(lo, mu, sigma)

# It's outside if it's not between
def normal_probability_outside(lo: float, hi: float, mu: float = 0,
                               sigma: float = 1) -> float:
  """The probability that an N(mu, sigma) is not between lo and hi."""
  return 1 - normal_probability_between(lo, hi, mu, sigma)

<Figure size 640x480 with 0 Axes>

# Multiple Comparison

# Confidence Interval

A confidence interval (CI) provides a range of plausible values for a population parameter, such as a mean.

If you compute a 95% CI for the mean weight of candy bars:

$$ \bar{x} \pm 1.96 \cdot \frac{s}{\sqrt{n}} $$

This gives a range such that:

If you repeated the sampling process infinitely many times, 95% of the intervals would contain the true mean μ.


* A CI does not say “there is a 95% chance μ is in this interval.”
* Instead: “This method captures μ in 95% of repeated experiments.”
* The CI width reflects uncertainty. Smaller width = more precision.

Suppose:
* Sample mean = 247g
    * s1 = [243.1, 250.9], s2 = [246.8, 254.2] ....  sn
    * Each of these is a CI from a sample
    * $I1, I2, I3, I4, ...$ from $s1, s2, s3, s4, ... $
    * Of all these intervals 95% will contain the true mean $\mu$ and 5% will miss. We never know which interval is correct, but th procedure succeeds 95% of the time
	* Standard deviation = 20g
	* n = 100

95% CI:

$247 \pm 1.96 \cdot \frac{20}{10} = 247 \pm 3.92$

That is:

[243.08,\; 250.92]

Because 250g is inside this interval → we fail to reject H₀ at α = 0.05.

CI ↔ Hypothesis Test

* A 95% CI and a two-sided hypothesis test at α = 0.05 give the same decision:
* If the CI excludes 250 → reject H₀
* If the CI includes 250 → fail to reject H₀


CI Assumptions:
* Independent observations (selected independent of the others)
* Accurate Data (non biased)
*

# p-Hacking

With some help, you can find the "significant" hypothesis if you test enough hypotheses against your dataset

To do good science, decide on your hypotheses before looking at the data.

In [8]:
# p-Hacking
import random

from typing import List
def run_experiment() -> List[bool]:
  """Flips a fair coin 1000 times, True = heads, False = tails"""
  return [random.random() < 0.5 for _ in range(1000)]

def reject_fairness(experiment: List[bool]) -> bool:
  """Using the 5% significance levels"""
  num_heads = len([flip for flip in experiment if flip])
  return num_heads < 469 or num_heads > 531

random.seed(0)

experiments = [run_experiment() for _ in range(1000)]
num_rejections = len([experiment
                      for experiment in experiments
                      if reject_fairness(experiment)])
assert num_rejections == 46

#A/B Testing

Check the book/Book-Github for details

# Bayesian Inference

Check the book/Book-Github for details